nni/mutable/mutable.py

```diff
@@ -501,3 +508,608 @@ class MutableSymbol(LabeledMutable, Symbol, MutableExpression):
+        sample_val = sample[self.label]
+        if sample_val not in self.values:
+            return SampleValidationError(f'{sample_val} not found in {self.values}')
+        return None
+
+    def extra_repr(self) -> str:
+        if len(self.values) <= 7:
+            return f'{self.values!r}, label={self.label!r}'
+        return '[' +  \
+            ', '.join(map(repr, self.values[:3])) + \
+            ', ..., ' + \
+            ', '.join(map(repr, self.values[-3:])) + \
+            f'], label={self.label!r}'
+
+    def freeze(self, sample: Sample) -> Any:
+        self.validate(sample)
+        return sample[self.label]
+
+    def __len__(self):
+        return len(self.values)
+
+    def default(self, memo: Sample | None = None) -> Choice:
+        """The default() of :class:`Discrete` is the first value unless default value is set.
+
+        See Also
+        --------
+        Mutable.default
+        """
+        memo = {} if memo is None else memo
+        err = self.contains(memo)
+        if isinstance(err, SampleMissingError):
+            if self.default_value is not MISSING:
+                memo[self.label] = self.default_value
+            else:
+                memo[self.label] = self.values[0]
+        rv = self.freeze(memo)
+        if self.default_value is not MISSING and rv != self.default_value:
+            raise ValueError(f'Default value is specified to be {self.default_value} but got {rv}. '
+                             f'Please check the default value of {self.label}.')
+        return rv
+
+    def random(self, memo: Sample | None = None, random_state: RandomState | None = None) -> Choice:
+        """Randomly sample a value from choices.
+        Distribution is respected if provided.
+
+        See Also
+        --------
+        Mutable.random
+        """
+        memo = {} if memo is None else memo
+        if random_state is None:
+            random_state = RandomState()
+        err = self.contains(memo)
+        if isinstance(err, SampleMissingError):
+            index = random_state.choice(len(self.values), p=self.distribution)
+            memo[self.label] = self.values[index]
+        return self.freeze(memo)
+
+    def grid(self, memo: Sample | None = None, granularity: int | None = None) -> Iterable[Choice]:
+        """Return also values as a grid. Sorted by distribution from most likely to least likely.
+
+        See Also
+        --------
+        Mutable.grid
+        """
+        memo = {} if memo is None else memo
+        err = self.contains(memo)
+
+        if isinstance(err, SampleMissingError):
+            if all(dis == self.distribution[0] for dis in self.distribution):
+                # uniform distribution
+                values_perm = self.values
+            else:
+                # More heavily-distributed items are put upfront.
+                indices = sorted(list(range(len(self.values))), key=lambda i: self.distribution[i], reverse=True)
```

Comment:
sorted() takes an iterable, so `list` is unnecessary here. (too many parentheses)